src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
@DanielRosenwasser But by exactly what mechanism would we sometimes just use the declared type and other times compute the narrowed type?

@JsonFreeman Yes, but we'd then have to introduce a distinction between *resolving* the type of an expression (in which case we could defer and ignore the RHS) and *checking* the type of an expression. We don't have such a distinction currently.